enhancements/network/ovn-hypershift-live-migration.md

```diff
@@ -0,0 +1,314 @@
+│ address:             │            │ address:             │    
+│  0a:58:0a:f4:00:01   │            │  0a:58:0a:f4:00:01   │
+│  10.244.0.8          │            │  10.244.0.8          │
+└──────────────────────┘            └──────────────────────┘
+
+┌─────────────────────────────────┐
+│ dhcp-options 1234               │
+│   lease_time: 3500              │
+│   router: 169.254.1.1           │
+│   dns_server: [kubedns]         │
+│   server_id: 169.254.1.1        │
+│   server_mac: c0:ff:ee:00:00:01 │
+└─────────────────────────────────┘
+```
+
+**Virt-launcher pod address:**
+The CNI will not set an address at virt-launcher pod netns, that address is
+assigned to the VM with the DHCP options from the LSP, this allows to use 
+kubevirt bridge binding with pod networking and still do live migration.
+
+#### IPAM
+
+The point to point routing feature allows an address to be running at a node different from
+the one "owning" the subnet the address is coming from. This will happen
+after VM live migration.
+
+One scenario for live migration is to shut down the node the VMs were migrated
+from, this means that the IPAM node subnet should go back to the pool but since
+the migrated VMs contains IPs from it those IPs should reserved in case the 
+subnet is assigned to a new node.
+
+On that case before assigning the subnet to the node the VMs ips need to be 
+reserved so they don't get assigned to new pods
+
+Another scenario is ovn-kubernetes pods restarting after live migration, on 
+that case ovnkube-master should discover to what IP pool the VM belongs 
+(using an annotation or searching for subnets) and reserve the address.
+
+#### Detecting hypershift VMs and live migration
+
+This enhancement is implemented as part of ovn-k pod's CNI and controllers so
+the feature detection mechanism have to be implemented as part of kubevirt
+virt-launcher pods.
+
+To skip CNI IP configuration and assign assign DHCP options to LSP the following
+label and annotation are check to ensure that only kubevirt pods with explicit configuration
+are affected:
+- `ovn.org/skip-ip-configuration-on-cni`
+- `kubevirt.io/vm`
+
+As part of this enhancement ovn-k should configure a point to point routing after 
+live migration to redirect it to the correct node and port, this feature is only
+activated if the pod is annotated with `kubevirt.io/allow-pod-bridge-network-live-migration`.
+
+During live migration there is source and target virt-launcher pod's with different names, the ovn-k pod's controller watch the following
+annotation and label at target pod `kubevirt.io/nodeName` and `kubevirt.io/migration-target-start-timestamp`:
+- The `kubevirt.io/nodeName` is set after the VM finishes live migrating or when it becomes ready.
+- The `kubevirt.io/migration-target-start-timestamp` is set when live migration has not finished but migration-target pod is ready to receive traffic (this happens at post-copy live migration, where migration is taking too long).
```

Comment:
I'm not sure I understand well enough what is happening here.

IIUC, point-to-point routing will **only** be configured for pods with the `kubevirt.io/allow-pod-bridge-network-live-migration` annotation, and only when the controller spots the `kubevirt.io/nodeName` annotation to be the "target" node, or when the `kubevirt.io/migration-target-start-timestamp` annotation is set ?

Reply:
Yep the `kubevirt.io/allow-pod-bridge-network-live-migration` is for the point to point routing al together and the nodeName and migration-target-start-timestamp is to create/update point to point routing (it signals when the live migrated target is ready)

Reply:
OK I see, thanks for clarifying. 

When do we "remove" the static route + policy ? I.e. what's the reconciliation criteria ?

Reply:
> OK I see, thanks for clarifying.
> 
> When do we "remove" the static route + policy ? I.e. what's the reconciliation criteria ?

Two cases:
- VM is removed
- VM is live migrated back to the node that owns its IP.

Reply:
OK. Please add this info to the doc then.

Reply:
done

Reply:
Do we also remove it when the pod is simply killed and moved to another host ? I.e rescheduled?

Reply:
> Do we also remove it when the pod is simply killed and moved to another host ? I.e rescheduled? Edit I guess that's covered under vm removed.

@martinkennelly, yes, whenever the VM is moved it check if ptp routing need to be created/deleted/modified and also the cleanup process always remove it.